nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -14,84 +14,158 @@
-      current_costmap_msg->metadata.size_x, current_costmap_msg->metadata.size_y,
-      current_costmap_msg->metadata.resolution,
-      current_costmap_msg->metadata.origin.position.x,
-      current_costmap_msg->metadata.origin.position.y);
+    std::lock_guard<std::mutex> lock(costmap_msg_mutex_);
+    costmap_msg_ = msg;
+  }
+  if (!costmap_received_) {
+    costmap_ = std::make_shared<Costmap2D>(
+      msg->metadata.size_x, msg->metadata.size_y,
+      msg->metadata.resolution, msg->metadata.origin.position.x,
+      msg->metadata.origin.position.y);
+
+    processCurrentCostmapMsg();
+
+    costmap_received_ = true;
   }
+}
+
+void CostmapSubscriber::costmapUpdateCallback(
+  const nav2_msgs::msg::CostmapUpdate::SharedPtr update_msg)
+{
+  if (costmap_received_) {
+    if (costmap_msg_ != nullptr) {
+      processCurrentCostmapMsg();
+    }
+
+    std::lock_guard<Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
+
+    auto map_cell_size_x = costmap_->getSizeInCellsX();
+    auto map_call_size_y = costmap_->getSizeInCellsY();
 
-  unsigned char * master_array = costmap_->getCharMap();
-  unsigned int index = 0;
-  for (unsigned int i = 0; i < current_costmap_msg->metadata.size_x; ++i) {
-    for (unsigned int j = 0; j < current_costmap_msg->metadata.size_y; ++j) {
-      master_array[index] = current_costmap_msg->data[index];
-      ++index;
+    if (map_cell_size_x < update_msg->x + update_msg->size_x ||
+      map_call_size_y < update_msg->y + update_msg->size_y)
+    {
+      RCLCPP_WARN(
+        logger_, "Update area outside of original map area. Costmap bounds: %d X %d, "
+        "Update origin: %d, %d  bounds: %d X %d", map_cell_size_x, map_call_size_y,
+        update_msg->x, update_msg->y, update_msg->size_x, update_msg->size_y);
+      return;
     }
+
+    // todo: check if frame is the same
+
+    auto master_array = costmap_->getCharMap();
+    // copy update msg row-wise
+    for (size_t y = 0; y < update_msg->size_y; ++y) {
+      auto starting_index_of_row_update_in_costmap = (y + update_msg->y) * map_cell_size_x +
+        update_msg->x;
+
+      std::copy_n(
+        update_msg->data.begin() + (y * update_msg->size_x),
+        update_msg->size_x, &master_array[starting_index_of_row_update_in_costmap]);
+    }
+  } else {
+    RCLCPP_WARN(logger_, "No costmap received.");
   }
 }
 
-void CostmapSubscriber::costmapCallback(const nav2_msgs::msg::Costmap::SharedPtr msg)
+void CostmapSubscriber::processCurrentCostmapMsg()
 {
-  std::atomic_store(&costmap_msg_, msg);
-  if (!costmap_received_) {
-    costmap_received_ = true;
+  std::scoped_lock lock(*(costmap_->getMutex()), costmap_msg_mutex_);
+  if (haveCostmapParametersChanged()) {
+    costmap_->resizeMap(
+      costmap_msg_->metadata.size_x, costmap_msg_->metadata.size_y,
+      costmap_msg_->metadata.resolution,
+      costmap_msg_->metadata.origin.position.x,
+      costmap_msg_->metadata.origin.position.y);
   }
+
+  auto master_array = costmap_->getCharMap();
```

Comment:
Ditto

Reply:
resolved with: 33bf135588a90c6db0f4745b1155de7e00941f2c

Please resolve this conversation if you can see these changes now.